### PR TITLE
Improve reduce function overlow handling

### PR DESCRIPTION
**Pull request by nickva**
Add `log` as another option for `query_config.reduce_limit` instead of `true`
or `false` only. This warns user without crashing the view.

Account for reduce functions code size when checking for overlflow. Code size
was previously implicitly included in input size, skewing the result of
overflow check.

Increase threshold for overflow check. Before it was only 200. This
should prevent false positives in case users have a large accumulator object to
collect various stats.

(PR-ing to discuss. Ok to postpone this after 2.0 release if not needed a 2.0 bug fix)

### CODE REVIEW

File: share/server/views.js [hunk 1]
```diff
@@ -32,13 +35,21 @@ var Views = (function() {
...
     // TODO make reduce_limit config into a number
     if (State.query_config && State.query_config.reduce_limit &&
-          reduce_length > 200 && ((reduce_length * 2) > State.line_length)) {
-      var reduce_preview = "Current output: '"+(reduce_line.substring(0,100) + "'... (first 100 of "+reduce_length+" bytes)");
-      throw(["error", 
-        "reduce_overflow_error", 
-        "Reduce output must shrink more rapidly: "+reduce_preview]);
+          reduce_length > 4096 && ((reduce_length * 2) > input_length)) {
+      var log_message = [
+          "Reduce output must shrink more repidly:",
```
Comment (tonysun83) [inline share/server/views.js]: rapidly?

Comment (nickva) [inline share/server/views.js]: Good find. Thanks, Tony!

**Comment by kxepal** [comment]
-1

Two problems:

1. Boolean `reduce_limit` flag eventually starts taking string value "log". Things should be consistent and behaviour should be predictable. Here is not the case.

 2. `reduce_line` is truncated with a reason.  For instance, to not blow the browser when it get the MiBs of data back. But here you also add `input_line` as  is which is much longer than `reduce_line` by the design. So things will be get more terrible.  With `reduce_limit = "log"` you can also eventually cause high IO activity since all these data will get dumped on log file no matter that debug logs are turned off.

**Comment by nickva** [comment]
@kxepal  It would be useful to have a logging / warning only value for this setting. Is it better to call the value "warning?" or have a new additional one "if enabled don't actually enable, but just log" setting. That might be more confusing... What did you have in mind, what's a better setting?

Sorry, didn't follow this part:

> "But here you also add input_line as is which is much longer than reduce_line

What do you mean by adding the input line? The code does the opposite, it subtracts the code size from the line length so it considers data only. Before it was broken the way it worked. So someone could have created a 2MB  function body and then they could create reduce values as big as 1MB.

In general this is a fix to get this code to be useful. Currently because of how restricted the size is, it is actually disabled in production. Periodically customers without realizing create large reduce values, for example by doing `return values`, so trying to fix it.

> With reduce_limit = "log" you can also eventually cause high IO activity since all these data will get dumped on log 

Log is supposed to be enabled for a while as a diagnostic setting. After a while it would be obvious if it works or not. Notice, the log output only includes the line lengths, specifically to not flood it with too much data.

**Comment by kxepal** [comment]
@nickva 
> have a new additional one "if enabled don't actually enable, but just log" setting. That might be more confusing... What did you have in mind, what's a better setting?

`reduce_limit_debug = true|false` sounds good.

For the rest: sorry, I shouldn't read the code at early morning. I didn't recognized that you return out lines length, not the lines itself. So all the further words are the waste. No strong -1, but config options still have to be different.

**Comment by kxepal** [comment]
@nickva 
Hmm....ok. I still don't like it, but cannot propose anything better and the 

> For example reduce reduce_limit_debug would only work with reduce_limit = false, which is not intuitive.

Is indeed more confusing, so the problem is not solved, but just moved to the else domain.

**Comment by wohali** [comment]
@nickva Looks like @kxepal changed his -1 to a -0(ish), and the build passes (I'm rerunning the one platform that had a transient failure but expect success). Would you care to merge this change?

I'd like to see a JIRA ticket for this as well, and a test case if it makes sense.

**Comment by nickva** [comment]
Thanks for taking a look @wohali  and @kxepal 

There is a JIRA issue already: https://issues.apache.org/jira/browse/COUCHDB-3046.

Also rebased on master.

**Pull request merged**